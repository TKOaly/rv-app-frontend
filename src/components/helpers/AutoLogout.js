import { debounce, throttle } from 'lodash';
import React, { useEffect } from 'react';

export const AutoLogout = ({logUserOut, timerMs = 30 * 1000}) => {
    useEffect(() => {
        const eventThrottleMs = 300;

        // Callback
        const logoutInactiveUser = () => {
            logUserOut();
        };

        const debouncedLogout = debounce(logoutInactiveUser, timerMs);

        // Fire initially
        debouncedLogout();


        const KeyDownEvent = () => {
            debouncedLogout();
        };

        const MouseMoveEvent = () => {
            debouncedLogout();
        };

        const throttledKeyDownEvent = throttle(KeyDownEvent, eventThrottleMs);
        const throttledMouseMoveEvent = throttle(MouseMoveEvent, eventThrottleMs);

        document.addEventListener(
            'keydown',
            throttledKeyDownEvent,
            { passive: true }
        );
        document.addEventListener('mousemove', throttledMouseMoveEvent
            , {passive: true});

        return () => {
            document.removeEventListener('keydown', throttledKeyDownEvent, {
                passive: true
            });
            document.removeEventListener('mousemove', throttledMouseMoveEvent, {
                passive: true
            });
            throttledKeyDownEvent.cancel();
            throttledMouseMoveEvent.cancel();
            debouncedLogout.cancel();
        };
    }, [logUserOut, timerMs]);

    return <React.Fragment />;
};
