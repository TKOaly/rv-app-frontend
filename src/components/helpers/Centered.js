import React from 'react';
import './styles/Centered.css';

const Centered = (props) => (
    <div className="centered" style={props.zIndex && { zIndex: props.zIndex }}>
        {props.children}
    </div>
);

export default Centered;
