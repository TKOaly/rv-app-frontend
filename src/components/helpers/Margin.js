import React from 'react';

const Margin = (props) => (
    <div
        className="margin"
        style={{
            margin: props.margin,
            display: props.inlineBlock ? 'inline-block' : 'block'
        }}
    >
        {props.children}
    </div>
);

export default Margin;
