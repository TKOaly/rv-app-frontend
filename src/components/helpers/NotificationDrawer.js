import './styles/NotificationDrawer.scss';
import { Fade } from '../animations/Animations';
import { TransitionGroup } from 'react-transition-group';
import ErrorNotification from '../notifications/ErrorNotification';
import PurchaseNotification from '../notifications/PurchaseNotification';
import React from 'react';
import SuccessNotification from '../notifications/SuccessNotification';

const NotificationDrawer = ({ notifications, purchases }) => (
    <div className="notificationDrawer">
        <TransitionGroup>
            {notifications &&
                notifications.length > 0 &&
                notifications.map((notification, id) =>
                    notification.messageType === 'SUCCESS' ? (
                        <Fade key={notification.id}>
                            <SuccessNotification message={notification.message} shadow />
                        </Fade>
                    ) : (
                        <Fade key={notification.id}>
                            <ErrorNotification message={notification.message} shadow />
                        </Fade>
                    )
                )}
            {purchases && purchases.length > 0 && (
                <Fade>
                    <PurchaseNotification shadow purchases={purchases} />
                </Fade>
            )}
        </TransitionGroup>
    </div>
);

export default NotificationDrawer;
