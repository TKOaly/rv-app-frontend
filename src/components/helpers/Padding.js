import React from 'react';

const Padding = (props) => (
    <div
        className="padding"
        style={{
            padding: props.padding,
            display: props.inlineBlock ? 'inline-block' : 'block'
        }}
    >
        {props.children}
    </div>
);

export default Padding;
