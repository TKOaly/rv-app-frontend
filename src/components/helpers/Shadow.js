import './styles/Shadow.css';
import React from 'react';

const Shadow = (props) => {
    return <div className="shadow" />;
};

export default Shadow;
