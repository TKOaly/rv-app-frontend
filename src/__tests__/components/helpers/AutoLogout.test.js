import { AutoLogout } from '../../../components/helpers/AutoLogout';
import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

describe('<AutoLogout />', () => {
    let addEventListenerMock, removeEventListenerMock;
    beforeEach(() => {
        addEventListenerMock = sinon.stub(document, 'addEventListener');
        removeEventListenerMock = sinon.stub(document, 'removeEventListener');
    });
    afterEach(() => {
        removeEventListenerMock.restore();
        addEventListenerMock.restore();
    });
    it('should match snapshot', () => {
        const cmpnt = renderer.create(<AutoLogout />).toJSON();
        expect(cmpnt).toMatchSnapshot();
    });
    it('should call addEventListener() twice when mounting', () => {
        expect(addEventListenerMock.notCalled).toBe(true);
        mount(<AutoLogout />);
        expect(addEventListenerMock.calledTwice).toBe(true);
    });
    it('should call removeEventListener() twice when unmounting', () => {
        expect(removeEventListenerMock.notCalled).toBe(true);
        const cmpnt = mount(<AutoLogout />);
        cmpnt.unmount();
        expect(removeEventListenerMock.calledTwice).toBe(true);
    });
    it('should call addEventListener() and removeEventListener() twice during the components lifetime', () => {
        expect(removeEventListenerMock.notCalled).toBe(true);
        expect(addEventListenerMock.notCalled).toBe(true);

        const cmpnt = mount(<AutoLogout timerMs={1000} />);
        cmpnt.unmount();

        expect(addEventListenerMock.calledTwice).toBe(true);
        expect(removeEventListenerMock.calledTwice).toBe(true);
    });
    it('should call logUserOut() after waiting for a period of time', () => {
        const clock = sinon.useFakeTimers();
        const logoutSpy = sinon.spy();

        expect(logoutSpy.notCalled).toBe(true);

        mount(<AutoLogout logUserOut={logoutSpy} timerMs={500} />);

        clock.tick(499);
        expect(logoutSpy.notCalled).toBe(true);

        clock.tick(1);
        expect(logoutSpy.callCount).toBe(1);

        clock.restore();
    });

    it('should not call logUserOut() after unmounting', () => {
        const clock = sinon.useFakeTimers();
        const logoutSpy = sinon.spy();

        expect(logoutSpy.notCalled).toBe(true);

        const cmpnt = mount(<AutoLogout logUserOut={logoutSpy} timerMs={500} />);

        clock.tick(450);

        expect(logoutSpy.notCalled).toBe(true);
        cmpnt.unmount();

        clock.tick(50);
        expect(logoutSpy.notCalled).toBe(true);

        clock.restore();
    });

    /*
    it('should reset timer when pressing a key', () => {
        removeEventListenerMock.restore();
        addEventListenerMock.restore();

        const clock = sinon.useFakeTimers();
        const logoutSpy = sinon.spy();

        const cmpnt = mount(<div className="wrap"><AutoLogout logUserOut={logoutSpy} timerMs={410} /></div>);

        clock.tick(400);

        expect(logoutSpy.notCalled).toBe(true);

        cmpnt.find('.wrap').simulate('keydown', { key: 'a' }); // This should add 410ms to the current timer

        clock.tick(50);

        expect(logoutSpy.notCalled).toBe(true);

        clock.tick(360);

        expect(logoutSpy.calledOnce).toBe(true);

        clock.restore();
    });*/
});
