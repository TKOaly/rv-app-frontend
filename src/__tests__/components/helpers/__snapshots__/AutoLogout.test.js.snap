// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoLogout /> should match snapshot 1`] = `null`;
